### ISSUE TITLE
Context resolveDataProvider doesn't generically handle typesystems with multiple package names

### ISSUE THREAD

**Issue by csandersdev** [DBCG/cql_engine#479]
`Context.resolveDataProvider(String packageName, boolean mustResolve)` assumes that the data provider / model resolver does resolution for classes sharing a single package name. It is clear that early versions of the HAPI FHIR typesystem did not follow this model as there is a hack in that method for handling `ca.uhn.fhir.model.dstu2` and `ca.uhn.fhir.model.primitive`. I'm now having the exact same issue with the IBM FHIR model objects. The resources and types are stored in different packages. 

I would like to be able to write a single data provider that can handle multiple package names. The simplest answer seems like it would be for the data provider / model resolver to advertise multiple package names (thus changing `String getPackageName()` to `List<String> getPackageName()`). I'm not super keen on changing the public interface though. Other ideas?

**Comment by JPercival** [comment]
We ran into this when patching a few gaps in the HAPI models as well which is why some of the jars include types in the org.hl7.fhir package.

https://github.com/DBCG/cql_engine/tree/master/engine.fhir/src/main/java/org/hl7/fhir/r4/model

I think the "real" fix is to include type information in the ELM such that the data providers can be resolved solely on the namespace declared in the ELM. @brynrhodes ?

**Comment by brynrhodes** [comment]
The ELM does include type information, the challenge is that the engine needs to be able to resolve a DataProvider based only on the name of the class, since we don't want to force providers to support particular interfaces on runtime objects.

I think it makes sense to introduce a `getPackageNames()` method and mark the `getPackageName()` deprecated, give it a few releases, then remove it.

**Comment by JPercival** [comment]
If we have namespace and name then it's not a problem, right? We can resolve the DataProvider according to the namespace, and then resolve the type from there.

**Comment by JPercival** [comment]
Here's an initial pass at what passing the namespace uri would look like. Didn't bother to get all the tests passing, just posting it here for discussion:

https://github.com/DBCG/cql_engine/compare/master...pass-namespace

**Comment by csandersdev** [comment]
@brynrhodes The solution @JPercival  provided seems sensible, but I don't claim to understand the finer points of the design. I'm open to whatever approach you guys think is best. I provided [a PR](480) for expanding the package names from a String to a List of Strings if you think that is preferable.

**Comment by JPercival** [comment]
Talked to Bryn about it. The issue is that while ELM CAN have type information the type information is optional, so it's not the case that it ALWAYS has type information. So the approach I showed won't always work. We talked through some potential workarounds, such as inferring all the type information ahead of time in a query-planning phase, but they are still a little way out. So given we don't always have type information, and given that we can't infer type information right now, and given that changing the API to `getPackageNames` covers the immediate need, we're going to do that.
